Delay Java compilation during pipelining

eed3si9n opened the pull request:
This is a follow-up to https://github.com/sbt/zinc/pull/744. (This actually makes pipelining usable)

![pipelining](https://user-images.githubusercontent.com/184683/81020232-58f43900-8e36-11ea-8239-3bacb0452a8b.png)

This implements an Analysis-aware `compileAllJava(...)`.

```scala
def compileAllJava(in: Inputs, logger: Logger): CompileResult = ...
```

How this is realized is more complicated than one might expect because we need all the wiring to get AnalysisCallback going to reuse the `Incremental` + `MixedAnalyzingCompiler` harness (This is so the *downstream* subprojects can get the incremental goodness).

**Note**: If incOptions pipelining is enabled, the build tool must call `compileAllJava(in, logger)` after _both_
1. All upstream subproject Scala and Java compilation has completed, and 
2. Scala *whole* `*.class` compilation has completed && hasModified.

eed3si9n committed:
Turn off pipelining by default

eed3si9n committed:
compileAllJava

This implements an Analysis-aware `compileAllJava(...)`.

```scala
def compileAllJava(in: Inputs, logger: Logger): CompileResult = ...
```

How this is realized is more complicated than one might expect because we need all the wiring to get AnalysisCallback going to reuse the `Incremental` + `MixedAnalyzingCompiler` harness (This is so the *downstream* subproject can get the incremental goodness).

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/843/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eed3si9n committed:
Turn off compileJava when pipelining is on

Note this requires build tools to call `incrementalCompiler.compileAllJava(in, log)` at the right timing.

eed3si9n commented on internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncHandler.scala:
```diff
@@ -533,22 +533,31 @@ case class ProjectStructure(
         // to avoid deletion of directories etc, we need to wait for the upstream to finish
         val f = for {
-          _ <- Future.traverse(dependsOnRef)(_.compile(i))
-          a <- futureAnalysis
+          pj <- wholeDeps
+          a0 <- futureScalaAnalysis
+          a <- futureJavaAnalysis(pj, a0)
         } yield a
```
This is the part that waits for upstream whole compilation and self Scala compilation for scripted tests.

eed3si9n committed:
Test with pipelining turned on

eed3si9n commented on zinc/src/main/scala/sbt/internal/inc/MixedAnalyzingCompiler.scala:
```diff
@@ -173,18 +181,25 @@ final class MixedAnalyzingCompiler(
+      compileScala()
+      if (scalaSrcs.nonEmpty) {
+        log.debug("done compiling Scala sources")
+      }
```
Java compilation is turned off when `incOptions.pipelining` is turned on.

eed3si9n commented:
@dwijnand

> But #843 with a `javaOnly` IncOption I think I'd be happy with.

`IncrementalCompiler#compile(in, log)` performs incremental compilation, and `IncrementalCompiler#compileAllJava(in, log)` performs non-incremental Java compilation and "analysis." Conceptually, we are splitting a subproject compilation into 2 parts (`scalac` and `javac`), and two form a logical `compile` with a complex timing.

Treating this as another `IncrementalCompiler#compile(in, log)` would not work because various pruning and invalidation logic would be different. In other words, they are not simply copy-pasted methods.

eed3si9n committed:
Add a unit test to test no-op non-compilation

dwijnand commented on build.sbt:
```diff
@@ -75,6 +75,8 @@ ThisBuild / publishTo := {
 ThisBuild / pomIncludeRepository := (_ => false) // drop repos other than Maven Central from POM
 ThisBuild / mimaPreviousArtifacts := Set.empty
+// limit the number of concurrent test so testQuick works
+Global / concurrentRestrictions += Tags.limit(Tags.Test, 4)
```
What's going on here?  I never use `testQuick`.

eed3si9n replied:
When I run `~testQuick` at the root level, I noticed that it's running 8 tests at a time, each vying for memory and CPU, and sometimes it would get stuck like it's deadlocked. It could also be plain memory leak somewhere. Project matrix makes it easy to have parallel subprojects. I think global 4 is a sweet spot for 2G of RAM I'm giving to this build.

dwijnand commented on zinc/src/sbt-test/source-dependencies/less-inter-inv-java/test:
```diff
@@ -6,4 +6,4 @@ $ copy-file changes/A2.java A.java
 # no further iteration, because APIs of directs don't change
 > run
-> checkIterations 2
+> checkIterations 1
```
Could you update the comment?  Same with where we disable pipelining (including source-dependencies/subproject-dependency) could you leave a note with some reasoning, please?

eed3si9n replied:
In general Zinc's scripted test are like do X1, X2, ..., `compile`, and see if `A` was recompiled as a result, or not.
Pipelining itself is ok, but the `compileAllJava(...)` at the end messes this up because all Java classes are invalidated. This _reduces_ the fidelity of our test because we're essentially over-compiling all the time. I turned off `pipelining` due to that limitation.

You or others can/should try to work out smaller sets of Java sources to compile that retains the same characteristic of incrementality (by tracking list of Java sources that needs recompilation in Analysis etc).

I'd have to look more into the `less-inter-inv-java`.

eed3si9n replied:
There was a regression in https://github.com/sbt/zinc/pull/744 for non-pipelined cases that's losing some Analysis data.

eed3si9n replied:
Regression is addressed in https://github.com/sbt/zinc/pull/843/commits/62e4432024f3a60e356a255f531d5822089d2393.

dwijnand replied:
I think that's the one that was breaking the "Created a Compilations with duplicate compilations" in `MCompilations`.

But I guess we'll see what CI thinks.

dwijnand reviewed (approved):
I'm not a fan of `incHandlerOpt.getOrElse(sys.error("incHandler was expected"))`, but it's a minor downside when compared to the rest of the change.  Very nice work.

I just had a couple of easy questions/requests.

eed3si9n committed:
Fix compilations getting lost in non-pipelined cases

eed3si9n committed:
Add comment
